README.md
# pypulsar

[![Testing](https://github.com/keenborder786/PyPulsar/actions/workflows/ci.yml/badge.svg)](https://github.com/keenborder786/PyPulsar/actions/workflows/ci.yml) [![CodeQL](https://github.com/keenborder786/PyPulsar/actions/workflows/github-code-scanning/codeql/badge.svg)](https://github.com/keenborder786/PyPulsar/actions/workflows/github-code-scanning/codeql) ![Python](https://img.shields.io/badge/python-3670A0?style=for-the-badge&logo=python&logoColor=ffdd54)


A Python Package to interact with Pulsar Cluster. Package implements the features which are missing in the official python client for pulsar.

## Installation

```bash
$ pip install pypulsar
```

## Usage

- TODO

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`pypulsar` was created by Keenborder786. It is licensed under the terms of the MIT license.

CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.0.1 (04/02/2023)

- First release of `pypulsar`!

tests/test_pypulsar.py
from pandas import DataFrame

from pypulsar import Bookie, Cluster


def test_cluster():
    pulsar_cluster = Cluster(False, "localhost", "8080")
    assert isinstance(pulsar_cluster.all_bookie_info(), DataFrame) == True
    assert isinstance(
        pulsar_cluster.get_rack_placement_bookies(), dict) == True
    assert isinstance(pulsar_cluster.get_broker_stats_allocator(
        "default"), dict) == True
    assert (
        isinstance(
            pulsar_cluster.get_pending_bookie_client_op_stats(), DataFrame) == True
        and pulsar_cluster.get_pending_bookie_client_op_stats().columns.tolist()
        == [
            "Type",
            "dataLedgerOpenOp",
            "dataLedgerCloseOp",
            "dataLedgerCreateOp",
            "dataLedgerDeleteOp",
            "cursorLedgerOpenOp",
            "cursorLedgerCloseOp",
            "cursorLedgerCreateOp",
            "cursorLedgerDeleteOp",
        ]
    ) or isinstance(pulsar_cluster.get_pending_bookie_client_op_stats(), dict) == True
    assert (
        isinstance(pulsar_cluster.get_broker_availability_report(
            "public", "default"), dict) == True
    )


def test_bookies():
    pulsar_bookie = Bookie(False, "localhost", "8080", "127.0.0.1:3181")
    assert isinstance(
        pulsar_bookie.delete_rack_placement_bookie(), dict) == True
    assert isinstance(pulsar_bookie.get_rack_placement_bookie(), dict) == True
    assert (
        isinstance(
            pulsar_bookie.update_rack_placement_info_bookie(
                {"hostname": "string", "rack": "string"}
            ),
            dict,
        )
        == True
    )


src/pypulsar/bookie.py
import json
from typing import Dict

import requests

from pypulsar.cluster import Cluster


class Bookie(Cluster):
    """

    The main class to interact with specific bookie on pulsar cluster

    """

    def __init__(self, secured: bool, webservice_url: str, webservice_port: str, bookie_id: str):
        """
        Parameters:
        ------------------------------

        secured(bool): http or https requests

        webservice_url(str): The web-service url for brokers to execute the api requests

        webservice_port(str): The web-service port for broker to execute the api requests

        bookie_id(str): The unique id to identify the bookie on pulsar cluster


        """
        super().__init__(secured, webservice_url, webservice_port)
        self.bookie_id = bookie_id

    def delete_rack_placement_bookie(self) -> Dict:
        """
        Removed the rack placement information for a specific bookie in the cluster

        """

        response = requests.delete(
            f"{self.request_type}://{self.webservice_url}:{self.webservice_port}/admin/v2/bookies/racks-info/{self.bookie_id}"
        )
        if response.status_code == 200:
            print("Rack placement information was deleted")
        elif response.status_code == 404:
            print(response.content.decode("utf-8"))
        elif response.status_code == 403:
            print("Don't have admin permissions")
        else:
            print(
                f"Request could not happen due to the following to error {response.status_code}")
        return {}

    def get_rack_placement_bookie(self) -> Dict:
        """
        Gets the rack placement information for a specific bookie in the cluster

        Returns:
        -------------------------------------

        Dict: Rack placement information for specific bookie in form of dictionary

        """
        response = requests.get(
            f"{self.request_type}://{self.webservice_url}:{self.webservice_port}/admin/v2/bookies/racks-info/{self.bookie_id}"
        )
        if response.status_code == 200:
            return json.loads(response.content.decode("utf-8"))
        elif response.status_code == 404:
            print(response.content.decode("utf-8"))
        elif response.status_code == 403:
            print("Don't have admin permissions")
        else:
            print(
                f"Request could not happen due to the following to error {response.status_code}")
        return {}

    def update_rack_placement_info_bookie(self, rack_info: Dict) -> Dict:
        """

        Updates the rack placement information for a specific bookie in the cluster

        Parameters:
        ----------------------------------------------

        rack_info(Dict): The rack placment information which you to use in the form:{"hostname": "string","rack": "string"}

        """
        response = requests.post(
            f"{self.request_type}://{self.webservice_url}:{self.webservice_port}/admin/v2/bookies/racks-info/{self.bookie_id}",
            headers={"Content-Type": "application/json"},
            data=json.dumps(rack_info),
        )
        if response.status_code == 200:
            print("The rack placement information was updated")
        elif response.status_code == 404:
            print(response.content.decode("utf-8"))
        elif response.status_code == 403:
            print("Don't have admin permissions")
        else:
            print(
                f"Request could not happen due to the following to error {response.status_code}")
        return {}


src/pypulsar/cluster.py
import json
from typing import Dict, Union

import pandas as pd
import requests
from pandas import DataFrame


class Cluster:
    """

    The class which act as an entry point to execute all cluster-wide operations


    """

    def __init__(self, secured: bool, webservice_url: str, webservice_port: str) -> None:
        """
        Parameters:
        ------------------------------

        secured(bool): http or https requests

        webservice_url(str): The web-service url for brokers to execute the api requests

        webservice_port(str): The web-service port for broker to execute the api requests

        """
        if not secured:
            self.request_type = "http"
        else:
            self.request_type = "https"
        self.webservice_url = webservice_url
        self.webservice_port = webservice_port

    def all_bookie_info(self) -> DataFrame:
        """

        Gets raw information for all the bookies in the cluster

        Returns
        ----------------------------

        DataFrame: All of the bookie information in form of dataframe.

        """
        response = requests.get(
            f"{self.request_type}://{self.webservice_url}:{self.webservice_port}/admin/v2/bookies/all"
        )
        if response.status_code == 200:
            final_response = json.loads(response.content)["bookies"]
            final_response_df = pd.DataFrame(final_response)
            return final_response_df
        elif response.status_code == 403:
            print("Don't have admin permission")
        else:
            print(
                f"Request could not happen due to the following to error {response.status_code}")
        return {}

    def get_rack_placement_bookies(self) -> Dict:
        """
        Gets the rack placement information for all the bookies in the cluster

        Returns:
        -------------------------------------

        Dict: Rack placement info for all bookie in form of dictionary

        """
        response = requests.get(
            f"{self.request_type}://{self.webservice_url}:{self.webservice_port}/admin/v2/bookies/racks-info"
        )
        if response.status_code == 200:
            final_response = json.loads(response.content)
            return final_response
        elif response.status_code == 403:
            print("Don't have admin permission")
        else:
            print(
                f"Request could not happen due to the following to error {response.status_code}")
        return {}

    def get_broker_stats_allocator(self, allocator: str) -> Dict:
        """

        Parameters:
        --------------------------------------

        allocator(str): Available allocators are 'default' and 'ml-cache'


        Get the stats for the Netty allocator. Available allocators are 'default' and 'ml-cache'

        Returns:
        --------------------------------------
        Dict: Stats for the Netty allocator in form of dictionary

        """
        response = requests.get(
            f"{self.request_type}://{self.webservice_url}:{self.webservice_port}/admin/v2/broker-stats/allocator-stats/{allocator}"
        )
        if response.status_code == 200:
            final_response = json.loads(response.content)
            return final_response
        elif response.status_code == 403:
            print("Don't have admin permission")
        else:
            print(
                f"Request could not happen due to the following to error {response.status_code}")

        return {}

    def get_pending_bookie_client_op_stats(self) -> Union[DataFrame, Dict]:
        """

        Get pending bookie client op stats by namespace

        Returns:
        --------------------------------------
        Dict: Pending bookie client op stats by namespace in form dataframe

        """
        response = requests.get(
            f"{self.request_type}://{self.webservice_url}:{self.webservice_port}/admin/v2/broker-stats/bookieops"
        )
        if response.status_code == 200:
            final_response = json.loads(response.content)
            keys = list(final_response.keys())
            final_dataframe_data = []
            for key in keys:
                topic_function_data = final_response[key]
                topics_functions = list(topic_function_data.keys())
                for topic_func in topics_functions:
                    instance = {}
                    instance["Type"] = topic_func
                    metric_data = topic_function_data[topic_func]
                    for key, value in metric_data.items():
                        instance[key] = value
                    final_dataframe_data.append(instance)
            df_final_response = pd.DataFrame(final_dataframe_data)
            return df_final_response
        elif response.status_code == 403:
            print("Don't have admin permission")
        else:
            print(
                f"Request could not happen due to the following to error {response.status_code}")

        return {}

    def get_broker_availability_report(self, tenant: str, namespace: str) -> Dict:
        """

        This API gives the current broker availability in percent, each resource percentage usage is calculated and
        thensum of all of the resource usage percent is called broker-resource-availability

        Parameters:
        --------------------
        tenant(str): The tenant for which you want availability report

        namespace(str): The namespace for which want availability report

        Returns:
        --------------------
        dict: The broker availability information in form of dictionary

        """

        response = requests.get(
            f"{self.request_type}://{self.webservice_url}:{self.webservice_port}/admin/v2/broker-stats/broker-resource-availability/{tenant}/{namespace}"
        )

        if response.status_code == 200:
            final_response = json.loads(response.content)
            return final_response
        elif response.status_code == 403:
            print("Don't have admin permission")
        else:
            print(
                f"Request could not happen due to the following to error {response.status_code}")

        return {}
